lib/rexml/source.rb

```diff
@@ -181,65 +151,40 @@ def initialize(arg, block_size=500, encoding=nil)
       end
     end
 
-    def scan(pattern, cons=false)
-      rv = super
-      # You'll notice that this next section is very similar to the same
-      # section in match(), but just a liiittle different.  This is
-      # because it is a touch faster to do it this way with scan()
-      # than the way match() does it; enough faster to warrant duplicating
-      # some code
-      if rv.size == 0
-        until @buffer =~ pattern or @source.nil?
-          begin
-            @buffer << readline
-          rescue Iconv::IllegalSequence
-            raise
-          rescue
-            @source = nil
-          end
-        end
-        rv = super
-      end
-      rv.taint if RUBY_VERSION < '2.7'
-      rv
-    end
-
     def read
       begin
-        @buffer << readline
+        @scanner.string = @scanner.rest + readline
```

Comment:
Can we use `@scanner << readline` here?

Reply:
Changing to `@scanner << readline` causes the following error in JRuby.

https://github.com/ruby/rexml/actions/runs/7434514894/job/20228750659#step:4:43
```
Error: test_rexml(REXMLTests::TestIssuezillaParsing):
  REXML::ParseException: No close tag for /issuezilla/issue[118]/activity[2]
  Line: -1
  Position: -1
  Last 80 unconsumed characters:
/Users/naitoh/ghq/github.com/naitoh/rexml/lib/rexml/parsers/treeparser.rb:28:in `parse'
/Users/naitoh/ghq/github.com/naitoh/rexml/lib/rexml/document.rb:448:in `build'
/Users/naitoh/ghq/github.com/naitoh/rexml/lib/rexml/document.rb:101:in `initialize'
org/jruby/RubyClass.java:904:in `new'
/Users/naitoh/ghq/github.com/naitoh/rexml/test/test_rexml_issuezilla.rb:8:in `block in test_rexml'
      5:     include Helper::Fixture
      6:     def test_rexml
      7:       doc = File.open(fixture_path("ofbiz-issues-full-177.xml")) do |f|
  =>  8:         REXML::Document.new(f)
      9:       end
     10:       ctr = 1
     11:       doc.root.each_element('//issue') do |issue|
org/jruby/RubyIO.java:1179:in `open'
/Users/naitoh/ghq/github.com/naitoh/rexml/test/test_rexml_issuezilla.rb:7:in `test_rexml'
org/jruby/RubyKernel.java:1310:in `catch'
org/jruby/RubyKernel.java:1305:in `catch'
org/jruby/RubyKernel.java:1310:in `catch'
org/jruby/RubyKernel.java:1305:in `catch'
```

I am not sure why the error is occurring, but I am thinking that https://github.com/ruby/strscan/issues/78 may be affected.

Reply:
It seems that the issue shows that `scanner.string = scanner.rest + XXX` has a problem but `scanner << XXX` doesn't have a problem. So it may not be related...

Reply:
https://github.com/ruby/strscan/issues/78 has been fixed. Could you try with the latest strscan?

Reply:
I tried, but it did not fix it...

Reply:
@kou 
I found the cause of this problem.

I was getting `Java::JavaLang::OutOfMemoryError: Java heap space` when trying to parse huge [XML](https://github.com/ruby/rexml/blob/master/test/data/ofbiz-issues-full-177.xml) in the DOM.


1. Rewrite `def read` as follows.
```ruby
    def read
      #begin
        @scanner << readline
      #rescue Exception, NameError
      #  @source = nil
      #end
    end
```

2. Rewrite `test/run.rb` as follows.
```ruby
diff --git a/test/run.rb b/test/run.rb
index 089318b..703aeb8 100755
--- a/test/run.rb
+++ b/test/run.rb
@@ -4,7 +4,7 @@ $VERBOSE = true
 
 base_dir = File.dirname(File.expand_path(__dir__))
 lib_dir = File.join(base_dir, "lib")
-test_dir = File.join(base_dir, "test")
+test_dir = "test/test_rexml_issuezilla.rb"
 
 $LOAD_PATH.unshift(lib_dir)
```

3. Run Test
```bash
$ bundle exec rake test
/Users/naitoh/.rbenv/versions/jruby-9.4.5.0/bin/jruby test/run.rb
/Users/naitoh/.rbenv/versions/jruby-9.4.5.0/lib/ruby/gems/shared/gems/power_assert-2.0.3/lib/power_assert.rb:8: warning: tracing (e.g. set_trace_func) will not capture all events without --debug flag
/Users/naitoh/.rbenv/versions/jruby-9.4.5.0/lib/ruby/gems/shared/gems/power_assert-2.0.3/lib/power_assert.rb:8: warning: tracing (e.g. set_trace_func) will not capture all events without --debug flag
Loaded suite REXMLTests::TestIssuezillaParsing
Started
Finished in 5.459958 seconds.
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
1 tests, 177 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
0.18 tests/s, 32.42 assertions/s
MacBook-Air:rexml naitoh$ bundle exec rake test
/Users/naitoh/.rbenv/versions/jruby-9.4.5.0/bin/jruby test/run.rb
/Users/naitoh/.rbenv/versions/jruby-9.4.5.0/lib/ruby/gems/shared/gems/power_assert-2.0.3/lib/power_assert.rb:8: warning: tracing (e.g. set_trace_func) will not capture all events without --debug flag
/Users/naitoh/.rbenv/versions/jruby-9.4.5.0/lib/ruby/gems/shared/gems/power_assert-2.0.3/lib/power_assert.rb:8: warning: tracing (e.g. set_trace_func) will not capture all events without --debug flag
Loaded suite REXMLTests::TestIssuezillaParsing
Started
E
============================================================================================================================================================================================================
Error: test_rexml(REXMLTests::TestIssuezillaParsing): Java::JavaLang::OutOfMemoryError: Java heap space
org.jruby.util.ByteList.ensure(ByteList.java:362)
org.jruby.util.io.EncodingUtils.strBufCat(EncodingUtils.java:1832)
org.jruby.util.io.EncodingUtils.strBufCat(EncodingUtils.java:1817)
org.jruby.util.io.EncodingUtils.encCrStrBufCat(EncodingUtils.java:1940)
org.jruby.util.io.EncodingUtils.encCrStrBufCat(EncodingUtils.java:1867)
org.jruby.RubyString.cat19(RubyString.java:1423)
org.jruby.RubyString.cat19(RubyString.java:1409)
org.jruby.RubyString.append(RubyString.java:2675)
org.jruby.ext.strscan.RubyStringScanner.concat(RubyStringScanner.java:207)
org.jruby.ext.strscan.RubyStringScanner$INVOKER$i$1$0$concat.call(RubyStringScanner$INVOKER$i$1$0$concat.gen)
org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:242)
Users.naitoh.ghq.github_dot_com.naitoh.rexml.lib.rexml.source.invokeOther2:\=\^\^(/Users/naitoh/ghq/github.com/naitoh/rexml/lib/rexml/source.rb:153)
Users.naitoh.ghq.github_dot_com.naitoh.rexml.lib.rexml.source.RUBY$method$read$0(/Users/naitoh/ghq/github.com/naitoh/rexml/lib/rexml/source.rb:153)
java.lang.invoke.LambdaForm$DMH/917568725.invokeStatic_L6_L(LambdaForm$DMH)
java.lang.invoke.LambdaForm$MH/1088104996.invokeExact_MT(LambdaForm$MH)
org.jruby.internal.runtime.methods.CompiledIRMethod.call(CompiledIRMethod.java:152)
org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:148)
org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:212)
org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:193)
Users.naitoh.ghq.github_dot_com.naitoh.rexml.lib.rexml.parsers.baseparser.invokeOther188:read(/Users/naitoh/ghq/github.com/naitoh/rexml/lib/rexml/parsers/baseparser.rb:351)
Users.naitoh.ghq.github_dot_com.naitoh.rexml.lib.rexml.parsers.baseparser.RUBY$method$pull_event$0(/Users/naitoh/ghq/github.com/naitoh/rexml/lib/rexml/parsers/baseparser.rb:351)
java.lang.invoke.LambdaForm$DMH/659059448.invokeStatic_L6_L(LambdaForm$DMH)
java.lang.invoke.LambdaForm$MH/1088104996.invokeExact_MT(LambdaForm$MH)
org.jruby.internal.runtime.methods.CompiledIRMethod.call(CompiledIRMethod.java:152)
org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:148)
org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:212)
org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:193)
org.jruby.runtime.callsite.CachingCallSite.fcall(CachingCallSite.java:199)
Users.naitoh.ghq.github_dot_com.naitoh.rexml.lib.rexml.parsers.baseparser.invokeOther3:pull_event(/Users/naitoh/ghq/github.com/naitoh/rexml/lib/rexml/parsers/baseparser.rb:183)
Users.naitoh.ghq.github_dot_com.naitoh.rexml.lib.rexml.parsers.baseparser.RUBY$method$pull$0(/Users/naitoh/ghq/github.com/naitoh/rexml/lib/rexml/parsers/baseparser.rb:183)
java.lang.invoke.LambdaForm$DMH/830863979.invokeStatic_L6_L(LambdaForm$DMH)
java.lang.invoke.LambdaForm$MH/1088104996.invokeExact_MT(LambdaForm$MH)
============================================================================================================================================================================================================
Finished in 7.390895 seconds.
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
1 tests, 0 assertions, 0 failures, 1 errors, 0 pendings, 0 omissions, 0 notifications
0% passed
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
0.14 tests/s, 0.00 assertions/s
rake aborted!
Command failed with status (1): [/Users/naitoh/.rbenv/versions/jruby-9.4.5.0/bin/jruby test/run.rb]
/Users/naitoh/ghq/github.com/naitoh/rexml/Rakefile:9:in `block in <main>'
org/jruby/ext/monitor/Monitor.java:82:in `synchronize'
/Users/naitoh/.rbenv/versions/jruby-9.4.5.0/bin/bundle:25:in `<main>'
Tasks: TOP => test
```

Reply:
OK. Could you add a comment why we use `@scanner.string = @scanner.rest + readline` instead of `@scanner << readline` here?

Reply:
OK.
Added comment.

Reply:
@naitoh Is there a bug in JRuby I should look at or is this expected given the size of the input?

Reply:
@headius Cc: @kou 
I think the problem is with JRuby's StringScanner, so I have created issue.

https://github.com/ruby/strscan/issues/83

Reply:
@naitoh Thank you! I will try to fix it today.

Reply:
ruby/strscan#83 is fixed and will be released soon!

Reply:
@kou
https://github.com/ruby/strscan/pull/84 has been merged into master and I confirmed that JRuby's `@scanner << readline` works.

Reply:
Thanks!